src/main/java/org/folio/services/iteration/IterationJobRunner.java

```diff
@@ -0,0 +1,201 @@
+import org.folio.services.kafka.InventoryProducerRecordBuilder;
+import org.folio.services.kafka.topic.KafkaTopic;
+
+public class IterationJobRunner {
+
+  public static final String ITERATION_JOB_ID_HEADER = "iteration-job-id";
+
+  private static final Logger log = LogManager.getLogger(IterationJobRunner.class);
+  private static final int POOL_SIZE = 2;
+  private static final int UPDATE_PUBLISHED_EVERY = 1000;
+  private static volatile WorkerExecutor workerExecutor;
+
+  private final Context vertxContext;
+  private final Map<String, String> okapiHeaders;
+  private final PostgresClientFuturized postgresClient;
+  private final IterationJobRepository repository;
+  private CommonDomainEventPublisher<Instance> eventPublisher;
+
+
+  public IterationJobRunner(Context vertxContext, Map<String, String> okapiHeaders) {
+    this(new PostgresClientFuturized(PgUtil.postgresClient(vertxContext, okapiHeaders)),
+      new IterationJobRepository(vertxContext, okapiHeaders),
+      vertxContext,
+      okapiHeaders);
+  }
+
+  public IterationJobRunner(PostgresClientFuturized postgresClient, IterationJobRepository repository,
+                          Context vertxContext, Map<String, String> okapiHeaders) {
+    this.vertxContext = vertxContext;
+    this.okapiHeaders = okapiHeaders;
+
+    this.postgresClient = postgresClient;
+    this.repository = repository;
+
+    initWorker(vertxContext);
+  }
+
+  public void startIteration(IterationJob job) {
+    eventPublisher = new CommonDomainEventPublisher<>(vertxContext, okapiHeaders,
+        KafkaTopic.forName(job.getJobParams().getTopicName(), tenantId(okapiHeaders), environmentName()));
+
+    workerExecutor.executeBlocking(
+      promise -> streamInstanceIds(new IterationContext(job))
+        .map(notUsed -> null)
+        .onComplete(promise))
+      .map(notUsed -> null);
+  }
+
+  private Future<Long> streamInstanceIds(IterationContext context) {
+    return postgresClient.startTx()
+      .map(context::withConnection)
+      .compose(this::selectInstanceIds)
+      .map(context::withStream)
+      .compose(this::processStream)
+      .onComplete(recordsPublished -> {
+        context.stream.close()
+          .onComplete(notUsed -> postgresClient.endTx(context.connection))
+          .onFailure(error -> log.warn("Unable to commit transaction", error));
+
+        if (recordsPublished.failed()) {
+          log.warn("Unable to iterate instances", recordsPublished.cause());
+
+          logFailedJob(context);
+        } else {
+          var published = recordsPublished.result();
+          log.info("Instance iteration completed: totalRecords = {}", published);
+
+          logReindexCompleted(published, context);
+        }
+      });
+  }
+
+  private Future<RowStream<Row>> selectInstanceIds(IterationContext ctx) {
+    return postgresClient.selectStream(ctx.connection,
+      "SELECT id FROM " + postgresClient.getFullTableName(INSTANCE_TABLE));
+  }
```

Comment:
Could this be part of the repository, allowing the details to be hidden from this class (and allow the table name to be private to the repository)?

Reply:
Updated. Thanks